File: db-tester-junit/src/main/java/io/github/seijikohara/dbtester/junit/jupiter/extension/DatabaseTestExtension.java
package io.github.seijikohara.dbtester.junit.jupiter.extension;

import io.github.seijikohara.dbtester.api.annotation.Expectation;
import io.github.seijikohara.dbtester.api.annotation.Preparation;
import io.github.seijikohara.dbtester.api.config.Configuration;
import io.github.seijikohara.dbtester.api.config.DataSourceRegistry;
import io.github.seijikohara.dbtester.internal.context.TestContext;
import io.github.seijikohara.dbtester.junit.jupiter.lifecycle.ExpectationVerifier;
import io.github.seijikohara.dbtester.junit.jupiter.lifecycle.PreparationExecutor;
import java.util.Optional;
import java.util.function.Supplier;
import org.junit.jupiter.api.extension.AfterEachCallback;
import org.junit.jupiter.api.extension.BeforeEachCallback;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.junit.jupiter.api.extension.ExtensionContext.Namespace;
import org.junit.jupiter.api.extension.ExtensionContext.Store;
import org.junit.jupiter.api.extension.ParameterContext;
import org.junit.jupiter.api.extension.ParameterResolver;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * JUnit Jupiter extension for database testing.
 *
 * <p>This extension processes {@link Preparation} and {@link Expectation} annotations to set up
 * test data before each test and verify database state after each test.
 *
 * <p>The extension performs three responsibilities:
 *
 * <ol>
 *   <li>Manages per-class state (configuration, {@link DataSourceRegistry}, lifecycle coordinator)
 *       using the {@link ExtensionContext} store.
 *   <li>Before each test, resolves {@link Preparation} declarations and executes the resulting
 *       datasets.
 *   <li>After each test, resolves {@link Expectation} declarations and validates the database
 *       contents.
 * </ol>
 *
 * @see Preparation
 * @see Expectation
 */
public class DatabaseTestExtension
    implements BeforeEachCallback, AfterEachCallback, ParameterResolver {

  /** Logger for tracking test execution progress and errors. */
  private static final Logger logger = LoggerFactory.getLogger(DatabaseTestExtension.class);

  /** Store key for the configuration instance. */
  private static final String STORE_KEY_CONFIGURATION = "configuration";

  /** Store key for the data source registry instance. */
  private static final String STORE_KEY_REGISTRY = "registry";

  /** Executor for handling test preparation phase. */
  private final PreparationExecutor preparationExecutor = new PreparationExecutor();

  /** Verifier for handling test expectation verification phase. */
  private final ExpectationVerifier expectationVerifier = new ExpectationVerifier();

  /**
   * Creates a database test extension with default configuration.
   *
   * <p>Configuration and {@link DataSourceRegistry} instances are created lazily per test class
   * using the JUnit {@link ExtensionContext} store mechanism. This ensures proper isolation between
   * test classes while enabling configuration sharing within a class hierarchy.
   */
  public DatabaseTestExtension() {
    // Configuration and registry are created lazily per test class via Store
  }

  /**
   * Gets or creates the DataSourceRegistry for the current test class.
   *
   * <p>The registry is stored in the root context's Store, making it available to all tests in the
   * class hierarchy. Tests should call this method in {@code @BeforeAll} to register data sources.
   *
   * @param extensionContext the extension context
   * @return the data source registry for this test class
   */
  public static DataSourceRegistry getRegistry(final ExtensionContext extensionContext) {
    return getOrCreate(
        extensionContext, STORE_KEY_REGISTRY, DataSourceRegistry.class, DataSourceRegistry::new);
  }

  /**
   * Stores a custom configuration for the current test class.
   *
   * <p>This method allows tests to customize the framework's behavior by providing a custom {@link
   * Configuration} instance. It must be called in {@code @BeforeAll} before any test execution to
   * ensure the configuration is available when needed.
   *
   * <p>If not called, the framework uses {@link Configuration#defaults()}.
   *
   * @param extensionContext the extension context
   * @param configuration the custom configuration to use for this test class
   */
  public static void setConfiguration(
      final ExtensionContext extensionContext, final Configuration configuration) {
    final var store = getClassScopedStore(extensionContext);
    store.put(STORE_KEY_CONFIGURATION, configuration);
  }

  /**
   * Called before each test method.
   *
   * <p>If the test method or class is annotated with {@link Preparation}, this callback executes
   * the preparation phase by loading and applying datasets to the database.
   *
   * @param context the current extension context
   */
  @Override
  public void beforeEach(final ExtensionContext context) {
    Optional.ofNullable(findPreparation(context))
        .ifPresent(
            preparation -> {
              final var testContext = createTestContext(context);
              logger.debug(
                  "Executing beforeEach for {}.{}()",
                  testContext.testClass().getSimpleName(),
                  testContext.testMethod().getName());
              preparationExecutor.execute(testContext, preparation);
            });
  }

  /**
   * Called after each test method.
   *
   * <p>If the test method or class is annotated with {@link Expectation}, this callback executes
   * the expectation phase by loading expected datasets and comparing them with the actual database
   * state.
   *
   * @param context the current extension context
   */
  @Override
  public void afterEach(final ExtensionContext context) {
    Optional.ofNullable(findExpectation(context))
        .ifPresent(
            expectation -> {
              final var testContext = createTestContext(context);
              logger.debug(
                  "Executing afterEach for {}.{}()",
                  testContext.testClass().getSimpleName(),
                  testContext.testMethod().getName());
              expectationVerifier.verify(testContext, expectation);
            });
  }

  /**
   * Determines if this extension can resolve the specified parameter.
   *
   * <p>This extension supports injecting {@link ExtensionContext} into test methods and lifecycle
   * methods (including {@code @BeforeAll} when using {@code @TestInstance(Lifecycle.PER_CLASS)}).
   *
   * @param parameterContext the context for the parameter
   * @param extensionContext the extension context
   * @return {@code true} if the parameter type is {@link ExtensionContext}
   */
  @Override
  public boolean supportsParameter(
      final ParameterContext parameterContext, final ExtensionContext extensionContext) {
    return parameterContext.getParameter().getType() == ExtensionContext.class;
  }

  /**
   * Resolves the parameter value.
   *
   * <p>Returns the {@link ExtensionContext} instance for the current test execution.
   *
   * @param parameterContext the context for the parameter
   * @param extensionContext the extension context
   * @return the extension context instance
   */
  @Override
  public Object resolveParameter(
      final ParameterContext parameterContext, final ExtensionContext extensionContext) {
    return extensionContext;
  }

  /**
   * Creates a {@link TestContext} from the JUnit {@link ExtensionContext}.
   *
   * <p>This method extracts the test class, test method, configuration, and registry from the
   * extension context and assembles them into a TestContext instance for use by lifecycle
   * executors.
   *
   * @param context the JUnit extension context providing test metadata
   * @return the test context containing all necessary information for test execution
   */
  private TestContext createTestContext(final ExtensionContext context) {
    final var testClass = context.getRequiredTestClass();
    final var testMethod = context.getRequiredTestMethod();
    final var configuration = getConfiguration(context);
    final var registry = getRegistry(context);

    return new TestContext(testClass, testMethod, configuration, registry);
  }

  /**
   * Finds the effective {@link Preparation} annotation for the current test.
   *
   * <p>Method-level annotations take precedence over class-level annotations. This method searches
   * first at the test method level, then falls back to the test class level if not found at the
   * method level.
   *
   * @param context the extension context providing access to test metadata
   * @return the preparation annotation if found at method or class level, or null if not present
   */
  private Preparation findPreparation(final ExtensionContext context) {
    final var method = context.getRequiredTestMethod();
    return Optional.ofNullable(method.getAnnotation(Preparation.class))
        .orElseGet(() -> context.getRequiredTestClass().getAnnotation(Preparation.class));
  }

  /**
   * Finds the effective {@link Expectation} annotation for the current test.
   *
   * <p>Method-level annotations take precedence over class-level annotations. This method searches
   * first at the test method level, then falls back to the test class level if not found at the
   * method level.
   *
   * @param context the extension context providing access to test metadata
   * @return the expectation annotation if found at method or class level, or null if not present
   */
  private Expectation findExpectation(final ExtensionContext context) {
    final var method = context.getRequiredTestMethod();
    return Optional.ofNullable(method.getAnnotation(Expectation.class))
        .orElseGet(() -> context.getRequiredTestClass().getAnnotation(Expectation.class));
  }

  /**
   * Gets or creates the Configuration for the current test class.
   *
   * <p>By default, uses {@link Configuration#defaults()}. Tests can customize configuration by
   * calling {@link #setConfiguration(ExtensionContext, Configuration)} before the first test runs.
   * The configuration is stored in the extension context store and shared across all tests in the
   * class hierarchy.
   *
   * @param extensionContext the extension context providing access to the configuration store
   * @return the configuration instance for this test class, either custom or default
   */
  private static Configuration getConfiguration(final ExtensionContext extensionContext) {
    return getOrCreate(
        extensionContext, STORE_KEY_CONFIGURATION, Configuration.class, Configuration::defaults);
  }

  /**
   * Generic get-or-create method for extension state management.
   *
   * <p>Retrieves an instance from the extension context store, creating and storing it if not
   * present. This pattern ensures lazy initialization and proper lifecycle management for per-class
   * extension state. The method is thread-safe through the extension context store mechanism.
   *
   * @param <T> the type of the instance to retrieve or create
   * @param extensionContext the extension context providing access to the store
   * @param key the storage key identifying the instance in the store
   * @param type the class of the instance for type-safe retrieval
   * @param factory the factory function to create new instances when not found in store
   * @return the existing instance from store, or newly created instance if not present
   */
  private static <T> T getOrCreate(
      final ExtensionContext extensionContext,
      final String key,
      final Class<T> type,
      final Supplier<T> factory) {
    final var store = getClassScopedStore(extensionContext);
    return Optional.ofNullable(store.get(key, type))
        .orElseGet(
            () -> {
              final var instance = factory.get();
              store.put(key, instance);
              return instance;
            });
  }

  /**
   * Returns the class-scoped store used to hold extension state for a specific test class.
   *
   * <p>For nested test classes, this method returns the store for the top-level test class to
   * ensure that state (such as data source registrations) is shared across all nested classes. The
   * store is created with a namespace unique to this extension and the top-level test class.
   *
   * @param extensionContext the current extension context providing access to test hierarchy
   * @return store scoped to the top-level test class, shared across nested test classes
   */
  private static Store getClassScopedStore(final ExtensionContext extensionContext) {
    final var topLevelTestClass = getTopLevelTestClass(extensionContext);
    final var namespace = Namespace.create(DatabaseTestExtension.class, topLevelTestClass);
    return extensionContext.getRoot().getStore(namespace);
  }

  /**
   * Finds the top-level test class by traversing up the context hierarchy.
   *
   * <p>For nested test classes, this method returns the outermost test class. For non-nested
   * classes, it returns the test class itself. This ensures configuration and state sharing across
   * the entire test class hierarchy.
   *
   * @param extensionContext the current extension context for hierarchy traversal
   * @return the top-level test class from the context hierarchy
   */
  private static Class<?> getTopLevelTestClass(final ExtensionContext extensionContext) {
    return findTopLevelContext(extensionContext).getRequiredTestClass();
  }

  /**
   * Recursively finds the top-level context by traversing up the parent hierarchy.
   *
   * <p>This method recursively navigates up the parent chain until it finds the root test context
   * with a test class. The recursion terminates when no parent with a test class exists.
   *
   * @param extensionContext the current extension context for recursive traversal
   * @return the top-level extension context at the root of the test hierarchy
   */
  private static ExtensionContext findTopLevelContext(final ExtensionContext extensionContext) {
    return extensionContext
        .getParent()
        .filter(parent -> parent.getTestClass().isPresent())
        .map(DatabaseTestExtension::findTopLevelContext)
        .orElse(extensionContext);
  }
}

File: db-tester-kotest/src/main/kotlin/io/github/seijikohara/dbtester/kotest/extension/DatabaseTestExtension.kt
package io.github.seijikohara.dbtester.kotest.extension

import io.github.seijikohara.dbtester.api.annotation.AnnotationUtils
import io.github.seijikohara.dbtester.api.annotation.DataSet
import io.github.seijikohara.dbtester.api.annotation.ExpectedDataSet
import io.github.seijikohara.dbtester.api.annotation.ExportDataSet
import io.github.seijikohara.dbtester.api.config.Configuration
import io.github.seijikohara.dbtester.api.config.DataSourceRegistry
import io.github.seijikohara.dbtester.api.context.TestContext
import io.github.seijikohara.dbtester.kotest.lifecycle.KotestExpectationVerifier
import io.github.seijikohara.dbtester.kotest.lifecycle.KotestExportExecutor
import io.github.seijikohara.dbtester.kotest.lifecycle.KotestPreparationExecutor
import io.kotest.core.extensions.TestCaseExtension
import io.kotest.core.spec.Spec
import io.kotest.core.test.TestCase
import io.kotest.engine.test.TestResult
import org.slf4j.LoggerFactory
import java.lang.reflect.Method
import kotlin.reflect.jvm.javaMethod

/**
 * Kotest extension for database testing.
 *
 * This extension processes [DataSet] and [ExpectedDataSet] annotations to set up
 * test data before each test and verify database state after each test.
 *
 * The extension performs two responsibilities:
 * 1. Before each test, resolves [DataSet] declarations and executes the resulting datasets.
 * 2. After each test, resolves [ExpectedDataSet] declarations and validates the database contents.
 *
 * **Usage with `@DatabaseTest` annotation and `DatabaseTestSupport` interface (recommended):**
 * ```kotlin
 * @DatabaseTest
 * class MyTest : AnnotationSpec(), DatabaseTestSupport {
 *
 *     override val dbTesterRegistry = DataSourceRegistry()
 *
 *     @BeforeAll
 *     fun setup() {
 *         dbTesterRegistry.registerDefault(createDataSource())
 *     }
 *
 *     @Test
 *     @DataSet
 *     @ExpectedDataSet
 *     fun `should verify database state`() {
 *         // test implementation
 *     }
 * }
 * ```
 *
 * **Usage with explicit extension registration:**
 * ```kotlin
 * class MyTest : AnnotationSpec() {
 *     private val registry = DataSourceRegistry()
 *
 *     init {
 *         extensions(DatabaseTestExtension(registryProvider = { registry }))
 *     }
 *
 *     @Test
 *     @DataSet
 *     @ExpectedDataSet
 *     fun `should verify database state`() {
 *         // test implementation
 *     }
 * }
 * ```
 *
 * When using the `@DatabaseTest` annotation, the spec class must implement
 * [DatabaseTestSupport] to provide the registry and optional configuration.
 *
 * @property registryProvider optional provider function that returns the [DataSourceRegistry].
 *     When null, the extension requires the spec to implement [DatabaseTestSupport].
 * @property configurationProvider optional provider function for custom [Configuration].
 *     When null, the extension uses [DatabaseTestSupport.dbTesterConfiguration] if available,
 *     falling back to [Configuration.defaults] if not found.
 * @see io.github.seijikohara.dbtester.kotest.annotation.DatabaseTest
 * @see DatabaseTestSupport
 * @see DataSet
 * @see ExpectedDataSet
 */
public class DatabaseTestExtension public constructor(
    private val registryProvider: (() -> DataSourceRegistry)? = null,
    private val configurationProvider: (() -> Configuration)? = null,
) : TestCaseExtension {
    // All primary-constructor parameters carry defaults so Kotlin synthesizes a
    // no-argument constructor. Kotest instantiates this extension reflectively
    // through @DatabaseTest / @ApplyExtension, which requires that no-arg constructor.

    /** Executor for the preparation phase. */
    private var preparationExecutor: KotestPreparationExecutor = KotestPreparationExecutor()

    /** Verifier for the expectation phase. */
    private var expectationVerifier: KotestExpectationVerifier = KotestExpectationVerifier()

    /** Executor for the export phase. */
    private var exportExecutor: KotestExportExecutor = KotestExportExecutor()

    /**
     * Creates an extension with injected lifecycle collaborators.
     *
     * <p>This constructor exists to supply test doubles for the preparation, verification, and
     * export phases. Production code uses the primary constructor.
     *
     * @param registryProvider optional provider for the [DataSourceRegistry], or null
     * @param configurationProvider optional provider for the [Configuration], or null
     * @param preparationExecutor the preparation executor
     * @param expectationVerifier the expectation verifier
     * @param exportExecutor the export executor
     */
    internal constructor(
        registryProvider: (() -> DataSourceRegistry)?,
        configurationProvider: (() -> Configuration)?,
        preparationExecutor: KotestPreparationExecutor,
        expectationVerifier: KotestExpectationVerifier,
        exportExecutor: KotestExportExecutor,
    ) : this(registryProvider, configurationProvider) {
        this.preparationExecutor = preparationExecutor
        this.expectationVerifier = expectationVerifier
        this.exportExecutor = exportExecutor
    }

    /** Companion object containing class-level constants and logger. */
    private companion object {
        private val logger = LoggerFactory.getLogger(DatabaseTestExtension::class.java)
    }

    /**
     * Intercepts test case execution to handle preparation and expectation phases.
     *
     * @param testCase the test case being executed
     * @param execute the function to execute the test case
     * @return the test result
     */
    override suspend fun intercept(
        testCase: TestCase,
        execute: suspend (TestCase) -> TestResult,
    ): TestResult =
        requireMethod(testCase).let { method ->
            val dataSet = findDataSet(testCase, method)
            val expectedDataSet = findExpectedDataSet(testCase, method)
            val exportDataSet = findExportDataSet(testCase, method)
            when {
                dataSet != null || expectedDataSet != null || exportDataSet != null -> {
                    executeWithAnnotations(testCase, method, execute, dataSet, expectedDataSet, exportDataSet)
                }

                else -> {
                    execute(testCase)
                }
            }
        }

    /**
     * Executes the test case with dataset, expected dataset, and/or export dataset handling.
     *
     * @param testCase the test case being executed
     * @param method the resolved test method
     * @param execute the function to execute the test case
     * @param dataSet the DataSet annotation, or null
     * @param expectedDataSet the ExpectedDataSet annotation, or null
     * @param exportDataSet the ExportDataSet annotation, or null
     * @return the test result
     */
    private suspend fun executeWithAnnotations(
        testCase: TestCase,
        method: Method,
        execute: suspend (TestCase) -> TestResult,
        dataSet: DataSet?,
        expectedDataSet: ExpectedDataSet?,
        exportDataSet: ExportDataSet?,
    ): TestResult =
        createTestContext(testCase, method).let { testContext ->
            dataSet?.also {
                logger.debug(
                    "Executing preparation for {}.{}()",
                    testContext.testClass().simpleName,
                    testContext.testMethod().name,
                )
                preparationExecutor.execute(testContext, it)
            }
            execute(testCase).also { result ->
                var testFailed = result !is TestResult.Success
                try {
                    if (result is TestResult.Success && expectedDataSet != null) {
                        logger.debug(
                            "Verifying expectation for {}.{}()",
                            testContext.testClass().simpleName,
                            testContext.testMethod().name,
                        )
                        expectationVerifier.verify(testContext, expectedDataSet)
                    }
                } catch (e: Throwable) {
                    testFailed = true
                    throw e
                } finally {
                    handleExportDataSet(testContext, exportDataSet, testFailed)
                }
            }
        }

    /**
     * Handles [ExportDataSet] execution in a finally-equivalent block.
     *
     * Export errors are caught and logged to prevent masking test or verification failures.
     *
     * @param testContext the test context
     * @param exportDataSet the ExportDataSet annotation, or null
     * @param testFailed whether the test execution or verification failed
     */
    private fun handleExportDataSet(
        testContext: TestContext,
        exportDataSet: ExportDataSet?,
        testFailed: Boolean,
    ) {
        if (exportDataSet == null) return
        if (exportDataSet.onFailureOnly && !testFailed) {
            logger.debug(
                "Skipping @ExportDataSet for {}.{}() because the test passed and onFailureOnly=true",
                testContext.testClass().simpleName,
                testContext.testMethod().name,
            )
            return
        }
        try {
            exportExecutor.export(testContext, exportDataSet)
        } catch (e: Exception) {
            logger.error(
                "Failed to export dataset for {}.{}(): {}",
                testContext.testClass().simpleName,
                testContext.testMethod().name,
                e.message,
                e,
            )
        }
    }

    /**
     * Creates a [TestContext] from the Kotest [TestCase].
     *
     * Resolves the [DataSourceRegistry] and [Configuration] using the following priority:
     * 1. Explicit provider functions passed to the constructor
     * 2. [DatabaseTestSupport] interface implementation
     * 3. Default values (empty registry causes error, default configuration)
     *
     * @param testCase the Kotest test case
     * @param method the resolved test method
     * @return the test context
     */
    private fun createTestContext(
        testCase: TestCase,
        method: Method,
    ): TestContext =
        TestContext(
            testCase.spec::class.java,
            method,
            resolveConfiguration(testCase.spec),
            resolveRegistry(testCase.spec),
        )

    /**
     * Resolves the [DataSourceRegistry] for the test.
     *
     * Resolution order:
     * 1. Explicit provider function (if provided)
     * 2. [DatabaseTestSupport.dbTesterRegistry] property
     * 3. Error (spec must implement DatabaseTestSupport)
     *
     * @param spec the spec instance
     * @return the resolved registry
     * @throws IllegalStateException if no registry is available
     */
    private fun resolveRegistry(spec: Spec): DataSourceRegistry =
        registryProvider?.invoke()
            ?: (spec as? DatabaseTestSupport)?.dbTesterRegistry
            ?: throw IllegalStateException(
                "Spec class '${spec::class.simpleName}' must implement DatabaseTestSupport " +
                    "to provide dbTesterRegistry, or use explicit registryProvider in " +
                    "DatabaseTestExtension constructor.",
            )

    /**
     * Resolves the [Configuration] for the test.
     *
     * Resolution order:
     * 1. Explicit provider function (if provided)
     * 2. [DatabaseTestSupport.dbTesterConfiguration] property
     * 3. Default configuration via [Configuration.defaults]
     *
     * @param spec the spec instance
     * @return the resolved configuration
     */
    private fun resolveConfiguration(spec: Spec): Configuration =
        configurationProvider?.invoke()
            ?: (spec as? DatabaseTestSupport)?.dbTesterConfiguration
            ?: Configuration.defaults()

    /**
     * Requires and returns the test method from the test case.
     *
     * For AnnotationSpec, the method name is derived from the test case name.
     * Kotlin preserves backtick-escaped method names verbatim at the JVM level
     * (spaces and all), so the test case name maps directly to the reflected
     * method name.
     *
     * @param testCase the test case
     * @return the resolved method
     * @throws IllegalStateException if method resolution fails
     */
    private fun requireMethod(testCase: TestCase): Method =
        testCase.spec::class.let { specClass ->
            specClass.members
                .firstOrNull { it.name == testCase.name.name }
                ?.let { member -> (member as? kotlin.reflect.KFunction<*>)?.javaMethod }
                ?: specClass.java.declaredMethods.firstOrNull { method ->
                    method.name == testCase.name.name
                }
        } ?: throw IllegalStateException(
            String.format(
                "Cannot resolve test method '%s' in class '%s'. " +
                    "DatabaseTestExtension requires AnnotationSpec style tests.",
                testCase.name.name,
                testCase.spec::class.java.name,
            ),
        )

    /**
     * Finds the effective [DataSet] annotation for the current test.
     *
     * Method-level annotations take precedence over class-level annotations.
     * Supports meta-annotation discovery through [AnnotationUtils].
     *
     * @param testCase the test case
     * @param method the resolved test method
     * @return the DataSet annotation if found, or null
     */
    private fun findDataSet(
        testCase: TestCase,
        method: Method,
    ): DataSet? =
        AnnotationUtils
            .findAnnotation(DataSet::class.java, method, testCase.spec::class.java)
            .orElse(null)

    /**
     * Finds the effective [ExpectedDataSet] annotation for the current test.
     *
     * Method-level annotations take precedence over class-level annotations.
     * Supports meta-annotation discovery through [AnnotationUtils].
     *
     * @param testCase the test case
     * @param method the resolved test method
     * @return the ExpectedDataSet annotation if found, or null
     */
    private fun findExpectedDataSet(
        testCase: TestCase,
        method: Method,
    ): ExpectedDataSet? =
        AnnotationUtils
            .findAnnotation(ExpectedDataSet::class.java, method, testCase.spec::class.java)
            .orElse(null)

    /**
     * Finds the effective [ExportDataSet] annotation for the current test.
     *
     * Method-level annotations take precedence over class-level annotations.
     * Supports meta-annotation discovery through [AnnotationUtils].
     *
     * @param testCase the test case
     * @param method the resolved test method
     * @return the ExportDataSet annotation if found, or null
     */
    private fun findExportDataSet(
        testCase: TestCase,
        method: Method,
    ): ExportDataSet? =
        AnnotationUtils
            .findAnnotation(ExportDataSet::class.java, method, testCase.spec::class.java)
            .orElse(null)
}
